File: Kalorie/Kalorie/Core/UseCases/SaveFoodItemPersonalPortionsUseCase.swift
//
//  SaveFoodItemPersonalPortionsUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 09.09.2026.
//

import Foundation

protocol SaveFoodItemPersonalPortionsUseCaseProtocol {
    func callAsFunction(barcode: String, portions: [FoodPortionDomain]) async throws
}

struct SaveFoodItemPersonalPortionsUseCase: SaveFoodItemPersonalPortionsUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(barcode: String, portions: [FoodPortionDomain]) async throws {
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        for portion in portions {
            if let error = FoodPortionValidation.validate(name: portion.name, grams: portion.grams) { throw error }
        }
        if let error = FoodPortionValidation.validate(portions: portions) { throw error }
        let dto = FoodItemPersonalPortionsDTO(id: barcode, portions: portions.map(FoodPortionDTO.init(portion:)))
        try await dataProvider.setAsync(dto, id: barcode, in: Constants.Firestore.foodItemPortions(userId: userId))
    }
}

#if DEBUG
struct SaveFoodItemPersonalPortionsUseCaseFake: SaveFoodItemPersonalPortionsUseCaseProtocol {

    // MARK: - Properties

    var shouldThrow = false

    // MARK: - Functions

    func callAsFunction(barcode: String, portions: [FoodPortionDomain]) async throws {
        if shouldThrow { throw URLError(.unknown) }
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/SaveFoodItemPersonalPortionsUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.FoodPortionDomain
import antoni.kalorie.core.models.FoodPortionValidation
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemPersonalPortionsDTO
import antoni.kalorie.core.networking.FoodPortionDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants

interface SaveFoodItemPersonalPortionsUseCaseProtocol {
    suspend operator fun invoke(barcode: String, portions: List<FoodPortionDomain>)
}

class SaveFoodItemPersonalPortionsUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : SaveFoodItemPersonalPortionsUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(barcode: String, portions: List<FoodPortionDomain>) {
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        for (portion in portions) {
            FoodPortionValidation.validate(name = portion.name, grams = portion.grams)?.let { throw it }
        }
        FoodPortionValidation.validate(portions)?.let { throw it }
        val dto = FoodItemPersonalPortionsDTO(id = barcode, portions = portions.map(::FoodPortionDTO))
        dataProvider.setAsync(dto, id = barcode, inCollection = Constants.Firestore.foodItemPortions(userId))
    }
}
